File: pixfold-d1/lib/domain/naming.dart
/// 命名结构求值：按组件序列生成建议名称，并做冲突/非法字符/大小写冲突检测。
library;

import 'models.dart';

/// 按命名结构对"当前顺序"（含人工调整）逐项生成建议名称。
/// [order] 为该集合的最终页序；imageIndex 组件取其在 order 中的位置。
List<NameProposal> buildProposals({
  required List<MockImage> order,
  required NamingScheme scheme,
  Map<String, int>? dirIndexMap,
  Map<String, String>? overrides,
}) {
  // 目录序号：集合内去重目录排序后 1-based 编号
  final dirs = order.map((m) => m.dir).toSet().toList()..sort();
  final dim = dirIndexMap ??
      {for (var i = 0; i < dirs.length; i++) dirs[i]: i + 1};

  final proposals = <NameProposal>[];
  for (var i = 0; i < order.length; i++) {
    final img = order[i];
    final parts = <String>[];
    final seps = <String>[];
    for (final comp in scheme.components) {
      if (!comp.enabled) continue;
      final part = _componentText(comp, img, i, dim, scheme);
      if (part.isEmpty) continue;
      parts.add(part);
      seps.add(comp.separatorBefore);
    }
    // 首个组件不加分隔符，其余按各自前置分隔符拼接
    var name = parts.isEmpty ? '' : parts.first;
    for (var k = 1; k < parts.length; k++) {
      name += seps[k] + parts[k];
    }
    // 扩展名策略
    switch (scheme.extPolicy) {
      case ExtPolicy.keep:
        name += '.${img.ext}';
      case ExtPolicy.lower:
        name += '.${img.ext.toLowerCase()}';
      case ExtPolicy.upper:
        name += '.${img.ext.toUpperCase()}';
      case ExtPolicy.strip:
        break;
    }
    if (scheme.sanitize) name = sanitizeName(name, scheme.replacement);
    final p = NameProposal(image: img, proposedName: name);
    p.overrideName = overrides?[img.id];
    proposals.add(p);
  }
  _attachWarnings(proposals, sanitize: scheme.sanitize);
  return proposals;
}

String _componentText(
  NameComponent comp,
  MockImage img,
  int position,
  Map<String, int> dirIndexMap,
  NamingScheme scheme,
) {
  String t;
  switch (comp.kind) {
    case NameComponentKind.prefix:
    case NameComponentKind.customText:
      t = comp.text;
    case NameComponentKind.rootDir:
      t = scheme.rootDirName;
    case NameComponentKind.relDir:
      t = img.dir;
    case NameComponentKind.dirIndex:
      final n = dirIndexMap[img.dir] ?? 1;
      t = n.toString().padLeft(scheme.indexPadding, '0');
    case NameComponentKind.imageIndex:
      final n = scheme.indexStart + position;
      t = n.toString().padLeft(scheme.indexPadding, '0');
    case NameComponentKind.origName:
      t = img.baseName;
  }
  switch (comp.casePolicy) {
    case CasePolicy.keep:
      break;
    case CasePolicy.lower:
      t = t.toLowerCase();
    case CasePolicy.upper:
      t = t.toUpperCase();
  }
  if (comp.trimSpaces) t = t.trim();
  return t;
}

/// 全工作区冲突检测：重名（完全相同）、大小写冲突（Windows 同目录不区分大小写）、非法字符、超长。
void _attachWarnings(List<NameProposal> proposals, {required bool sanitize}) {
  final byName = <String, List<NameProposal>>{};
  for (final p in proposals) {
    byName.putIfAbsent(p.finalName, () => []).add(p);
    if (p.isOverridden) {
      p.warnings.add(const ProposalWarning(ProposalWarningType.overridden, '人工覆盖'));
    }
  }
  for (final p in proposals) {
    if (!sanitize && containsIllegalChar(p.finalName)) {
      p.warnings.add(const ProposalWarning(
          ProposalWarningType.illegalChar, '包含非法字符 /\\:*?"<>|'));
    }
    if (p.finalName.length > 180) {
      p.warnings.add(const ProposalWarning(ProposalWarningType.tooLong, '名称过长（>180 字符）'));
    }
    if ((byName[p.finalName]?.length ?? 0) > 1) {
      p.warnings.add(ProposalWarning(
          ProposalWarningType.duplicate, '目标名称重复（${byName[p.finalName]!.length} 项同名）'));
    }
  }
  // 大小写冲突：不同名但 lower 后相同
  final byLower = <String, List<NameProposal>>{};
  for (final p in proposals) {
    byLower.putIfAbsent(p.finalName.toLowerCase(), () => []).add(p);
  }
  for (final group in byLower.values) {
    if (group.length < 2) continue;
    final names = group.map((p) => p.finalName).toSet();
    if (names.length < group.length) continue; // 已按重名报过
    for (final p in group) {
      p.warnings.add(const ProposalWarning(
          ProposalWarningType.caseCollision, 'Windows 下大小写冲突（同名不同大小写）'));
    }
  }
}

/// 生成重命名执行计划：冲突项默认跳过（供 UI 选择策略后调整）。
List<RenameOp> buildRenamePlan(List<NameProposal> proposals) {
  return proposals.map((p) {
    if (p.hasConflict) {
      return RenameOp(
        proposal: p,
        status: 'skipped',
        reason: p.warnings
            .firstWhere((w) =>
                w.type == ProposalWarningType.duplicate ||
                w.type == ProposalWarningType.illegalChar)
            .message,
      );
    }
    return RenameOp(proposal: p, status: 'ok');
  }).toList();
}

File: pixfold-d1/domain/src/main/kotlin/com/pixfold/d1/domain/model/Naming.kt
package com.pixfold.d1.domain.model

/**
 * 命名结构与逐项覆盖(规格 §4.4 / §6)。
 *
 * 关键设计:**覆盖(override)不放在 [NamingScheme] 里**,而是独立的
 * `Map<SourceItem.id, 完整文件名>`(见 §6.3)。改组件/分隔符/补零只影响
 * `proposedName`,被覆盖项逐字不变 —— 这是"改结构后覆盖不丢失"的机制本身,
 * 而不是靠 UI 小心维护。存储位置在 `NamingState`,不在本文件。
 */

/** 非法字符集:9 个 Windows 非法字符(规格 §7)。控制字符(码点 < 0x20)另行判定。 */
const val ILLEGAL_NAME_CHARS = "\\/:*?\"<>|"

/** 超长阈值:UTF-16 码元数,中文按 1 计(规格 §7)。 */
const val TOO_LONG_THRESHOLD = 180

enum class NameComponentKind(val label: String) {
    Prefix("自定义前缀"),
    RootDir("根目录名"),
    RelDir("相对目录片段"),
    DirIndex("目录序号"),
    ImageIndex("图片序号"),
    OrigName("原文件名"),
    CustomText("自定义文本"),
}

enum class CasePolicy(val label: String) {
    Keep("保持原样"),
    Lower("转小写"),
    Upper("转大写"),
}

enum class ExtPolicy(val label: String) {
    Keep("保留原样"),
    Lower("统一小写"),
    Upper("统一大写"),
    Strip("去除扩展名"),
}

/**
 * 命名组件。
 * @param separatorBefore 本组件**前置**分隔符;首个实际输出的组件不输出分隔符(§6.1)
 * @param text 仅 [NameComponentKind.Prefix] / [NameComponentKind.CustomText] 使用
 */
data class NameComponent(
    val kind: NameComponentKind,
    val enabled: Boolean = true,
    val separatorBefore: String = "_",
    val casePolicy: CasePolicy = CasePolicy.Keep,
    val trimSpaces: Boolean = true,
    val text: String = "",
)

data class NamingScheme(
    val rootDirName: String,
    val components: List<NameComponent> = DEFAULT_COMPONENTS,
    val indexStart: Int = 1,
    val indexPadding: Int = 3,
    val extPolicy: ExtPolicy = ExtPolicy.Lower,
    val sanitize: Boolean = true,
    val replacement: String = "_",
) {
    companion object {
        /** 默认输出 `<根目录名>_<三位序号>.<小写扩展名>`(规格 §4.4)。 */
        val DEFAULT_COMPONENTS = listOf(
            NameComponent(NameComponentKind.RootDir),
            NameComponent(NameComponentKind.ImageIndex),
            NameComponent(NameComponentKind.OrigName, enabled = false),
        )
    }
}

enum class ProposalWarningType(val label: String) {
    Duplicate("重名"),
    IllegalChar("非法字符"),
    CaseCollision("大小写冲突"),
    TooLong("超长"),
    Overridden("人工覆盖"),
}

data class ProposalWarning(val type: ProposalWarningType, val message: String)

/**
 * 单张图片的命名建议。
 *
 * [finalName] 优先取 [overrideName] —— 覆盖项逐字不变(§6.3)。
 */
data class NameProposal(
    val image: SourceItem,
    val proposedName: String,
    val overrideName: String? = null,
    val warnings: List<ProposalWarning> = emptyList(),
) {
    val finalName: String get() = overrideName ?: proposedName
    val isOverridden: Boolean get() = overrideName != null

    /** 会导致改名计划**跳过**的冲突:仅 Duplicate 与 IllegalChar(§7)。 */
    val hasConflict: Boolean
        get() = warnings.any {
            it.type == ProposalWarningType.Duplicate || it.type == ProposalWarningType.IllegalChar
        }

    /** 仅警告、不跳过计划的类型(CaseCollision / TooLong / Overridden)。 */
    val hasWarningOnly: Boolean
        get() = warnings.any { it.type == ProposalWarningType.CaseCollision || it.type == ProposalWarningType.TooLong }
}

/**
 * 命名批次状态(规格 §6.3)。
 *
 * [overrides] 键 = `SourceItem.id`,值 = 用户输入的完整文件名;**空值即移除**(清空输入框 = 撤销覆盖)。
 */
data class NamingState(
    val scheme: NamingScheme,
    val overrides: Map<String, String> = emptyMap(),
)
